consensus/hotstuff/committees/consensus_committee.go

```diff
@@ -236,6 +296,92 @@ func (c *Consensus) DKG(view uint64) (hotstuff.DKG, error) {
+// the protocol state. When we observe a new epoch being committed, we compute
+// the leader selection and cache static info for the epoch. When we observe
+// epoch emergency fallback being triggered, we inject a fallback epoch.
+func (c *Consensus) handleProtocolEvents(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case epoch := <-c.committedEpochsCh:
+			_, err := c.prepareEpoch(epoch)
+			if err != nil {
+				ctx.Throw(err)
+			}
+		case <-c.epochEmergencyFallback:
+			err := c.onEpochEmergencyFallbackTriggered()
+			if err != nil {
+				ctx.Throw(err)
+			}
+		}
+	}
+}
+
+// EpochCommittedPhaseStarted passes the protocol event to the worker thread.
+func (c *Consensus) EpochCommittedPhaseStarted(_ uint64, first *flow.Header) {
+	nextEpoch := c.state.AtBlockID(first.ID()).Epochs().Next()
+	c.committedEpochsCh <- nextEpoch
+}
+
+// EpochEmergencyFallbackTriggered passes the protocol event to the worker thread.
+func (c *Consensus) EpochEmergencyFallbackTriggered() {
+	c.epochEmergencyFallback <- struct{}{}
+}
+
+// onEpochEmergencyFallbackTriggered handles the protocol event for emergency epoch
+// fallback mode being triggered. When this occurs, we inject a fallback epoch
+// to the committee which extends the current epoch.
+// This method must also be called on initialization, if emergency fallback mode
+// was triggered in the past.
+// No errors are expected during normal operation.
+func (c *Consensus) onEpochEmergencyFallbackTriggered() (retErr error) {
+
+	// we respond to epoch fallback being triggered at most once, therefore
+	// the core logic is wrapped in sync.Once
+	// although it is only valid for epoch fallback to be triggered once per spork,
+	// we must account for repeated delivery of protocol events.
+	c.handleEpochFallbackOnce.Do(func() {
+		currentEpochCounter, err := c.state.Final().Epochs().Current().Counter()
+		if err != nil {
+			retErr = fmt.Errorf("could not get current epoch counter: %w", err)
+			return
+		}
+
+		c.mu.RLock()
+		// sanity check: current epoch must be cached already
+		currentEpoch, ok := c.epochs[currentEpochCounter]
+		c.mu.RUnlock()
+		if !ok {
+			retErr = fmt.Errorf("epoch fallback: could not find current epoch (counter=%d) info", currentEpochCounter)
+			return
+		}
+		// sanity check: next epoch must never be committed, therefore must not be cached
+		c.mu.RLock()
+		_, ok = c.epochs[currentEpochCounter+1]
+		c.mu.RUnlock()
+		if ok {
+			retErr = fmt.Errorf("epoch fallback: next epoch (counter=%d) is cached contrary to expectation", currentEpochCounter+1)
+			return
+		}
+
+		fallbackEpoch, err := newEmergencyFallbackEpoch(currentEpoch)
+		if err != nil {
+			retErr = fmt.Errorf("could not construct fallback epoch: %w", err)
+			return
+		}
+
+		// cache the epoch info
+		c.mu.Lock()
+		c.epochs[currentEpochCounter+1] = fallbackEpoch
+		c.mu.Unlock()
+	})
+	return
```

Comment:
do you need this extra return?

Reply:
Yeah, it's a compile error otherwise. Since the function has a return value - it's just named and set elsewhere